adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,285 @@
+				Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		secure := int8(1)
+		imp.Secure = &secure
+		imp.TagID = placementID
+	}
+	return request.Imp, nil
+}
+
+func getStoredRequestImpID(imp *openrtb.Imp) (string, error) {
+	var impExt map[string]json.RawMessage
+
+	err := json.Unmarshal(imp.Ext, &impExt)
+
+	if err != nil {
+		return "", fmt.Errorf("Error while decoding ext because: %s", err)
+	}
+
+	rawPrebidExt, ok := impExt[openrtb_ext.PrebidExtKey]
+
+	if !ok {
+		return "", fmt.Errorf("ext.prebid is null")
+	}
+
+	var prebidExt openrtb_ext.ExtImpPrebid
+
+	err = json.Unmarshal(rawPrebidExt, &prebidExt)
+
+	if err != nil {
+		return "", fmt.Errorf("cannot decoding ext.prebid because: %s", err)
+	}
+
+	if prebidExt.StoredRequest == nil {
+		return "", fmt.Errorf("ext.prebid.storedrequest is null")
+	}
+
+	return prebidExt.StoredRequest.ID, nil
+}
+
+func setSourceExt() (json.RawMessage, error) {
+	siteExt, err := json.Marshal(unicornSourceExt{
+		Stype:  "prebid_server_uncn",
+		Bidder: "unicorn",
+	})
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding source.ext, err: %s", err),
+		}
+	}
+	return siteExt, nil
+}
+
+func setExt(request *openrtb.BidRequest) (json.RawMessage, error) {
+	accountID, err := jsonparser.GetInt(request.Imp[0].Ext, "bidder", "accountId")
+	if err != nil {
+		accountID = 0
+	}
+	var decodedExt *unicornExt
+	err = json.Unmarshal(request.Ext, &decodedExt)
+	if err != nil {
+		decodedExt = &unicornExt{
+			Prebid: nil,
+		}
+	}
+	decodedExt.AccountID = accountID
+
+	ext, err := json.Marshal(decodedExt)
+	if err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error while encoding ext, err: %s", err),
+		}
+	}
+	return ext, nil
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *UnicornAdapter) MakeBids(request *openrtb.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected http status code: 400",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected http status code: %d", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			var bidType openrtb_ext.BidType
+			for _, imp := range request.Imp {
+				if imp.ID == bid.ImpID {
+					if imp.Banner != nil {
```

Comment:
This approach for determining bid type is an anti-pattern. Does your bidding server return any extra information you can use instead?

Reply:
I don't have the information from my bidding server need to determine the BidType, so I took this approach that other bidder adapters are doing.

By the way, most of the bidder adapters do not follow this mapping rule. What do you think?

Reply:
It's not a deal breaker. I always like to ask because publisher ad rendering may not work correctly if the bid type is incorrect. Please just try to match the behavior of your bidding server the best you can. For example, if you receive an imp with both banner and native sections defined, will your server prefer native? .. as this code would currently do.

Reply:
Certainly, there is a problem with the current conditional branching. understood.

Changed from "if" to "else if".